### PR TITLE
Change cert serial number generation

### PR DESCRIPTION
**Pull request by marcosy**
- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
Certificates serial numbers generation

**Description of change**
Currently, the certificate's serial numbers are generated sequentially. This PR makes those numbers to be generated randomly according to the CA/B forum considerations:

_(Section 7.1)_

> Effective September 30, 2016, CAs SHALL generate non-sequential Certificate serial numbers greater than zero (0) containing at least 64 bits of output from a CSPRNG

**Which issue this PR fixes**
Fixes: #962

### CODE REVIEW

File: pkg/common/x509util/serialnumber.go [hunk 1]
```diff
@@ -1,23 +1,28 @@
...
-}
-
-type serialNumber struct {
-	next int64
-}
+// NewSerialNumber creates a random certificate serial number according to CA/Browser forum spec
+// Section 7.1:
+//   "Effective September 30, 2016, CAs SHALL generate non-sequential Certificate serial numbers greater than
+//   zero (0) containing at least 64 bits of output from a CSPRNG"
+func NewSerialNumber() (*big.Int, error) {
```
Comment (azdagron) [inline pkg/common/x509util/serialnumber.go]: Nice!

File: pkg/common/x509util/serialnumber.go [hunk 2]
```diff
@@ -1,23 +1,28 @@
...
+	// Creates random integer in range [0,MaxUint64)
+	s, err := rand.Int(rand.Reader, maxUInt64)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create random number: %v", err)
+	}
 
-func (m *serialNumber) NextNumber(ctx context.Context) (*big.Int, error) {
-	return big.NewInt(atomic.AddInt64(&m.next, 1)), nil
+	// Adds 1 to return serial number [1,MaxUint64]
+	s.Add(s, big.NewInt(1))
```
Comment (azdagron) [inline pkg/common/x509util/serialnumber.go]: nit: return the value from Add directly (then people don't have to look up if Add mutates `s`, which it does, but isn't directly clear)
```
return s.Add(s, one)
```

File: pkg/common/x509util/serialnumber.go [hunk 3]
```diff
@@ -1,23 +1,28 @@
...
-	next int64
-}
+// NewSerialNumber creates a random certificate serial number according to CA/Browser forum spec
+// Section 7.1:
+//   "Effective September 30, 2016, CAs SHALL generate non-sequential Certificate serial numbers greater than
+//   zero (0) containing at least 64 bits of output from a CSPRNG"
+func NewSerialNumber() (*big.Int, error) {
+	// Get MaxUint64 as big.Int
+	maxUInt64 := new(big.Int)
+	maxUInt64 = maxUInt64.SetUint64(math.MaxUint64)
```
Comment (azdagron) [inline pkg/common/x509util/serialnumber.go]: nit: use package level unexported vars for these "const" values
```
var (
    maxUint64 = new(big.Int).SetUint64(math.MaxUint64)
    one = big.NewInt(1)
)
```
This is more for readability than performance (it keeps NewSerialNumber() small)

File: pkg/common/x509svid/upstreamca.go [hunk 4]
```diff
@@ -84,7 +80,7 @@ func (ca *UpstreamCA) SignCSR(ctx context.Context, csrDER []byte) (*x509.Certifi
 			x509.KeyUsageCertSign |
 			x509.KeyUsageCRLSign,
 		BasicConstraintsValid: true,
-		IsCA:                  true,
+		IsCA: true,
```
Comment (evan2645) [inline pkg/common/x509svid/upstreamca.go]: Is this a go fmt change that got reverted somehow?

File: pkg/common/x509util/serialnumber.go [hunk 5]
```diff
@@ -1,23 +1,28 @@
...
-func NewSerialNumber() SerialNumber {
-	return &serialNumber{}
-}
+// NewSerialNumber creates a random certificate serial number according to CA/Browser forum spec
+// Section 7.1:
+//   "Effective September 30, 2016, CAs SHALL generate non-sequential Certificate serial numbers greater than
+//   zero (0) containing at least 64 bits of output from a CSPRNG"
+func NewSerialNumber() (*big.Int, error) {
+	// Creates random integer in range [0,MaxUint64)
+	s, err := rand.Int(rand.Reader, maxUint64)
```
Comment (evan2645) [inline pkg/common/x509util/serialnumber.go]: CA/B forum mandates 64 bit minimum, RFC 5280 mandates 160 bit maximum:
```
   Given the uniqueness requirements above, serial numbers can be
   expected to contain long integers.  Certificate users MUST be able to
   handle serialNumber values up to 20 octets.  Conforming CAs MUST NOT
   use serialNumber values longer than 20 octets.
```

Perhaps we should choose something bigger than 64 bits? 128?

Comment (marcosy) [reply pkg/common/x509util/serialnumber.go]: Sounds good! I'll change it :+1:

File: pkg/common/x509svid/upstreamca.go (hunk 4 above)
Comment (marcosy) [reply pkg/common/x509svid/upstreamca.go]: Hmm... :thinking: It looks like when go fmt is called from my IDE on save, this line gets changed... I'll revert it.

File: pkg/common/x509util/serialnumber.go [hunk 6]
```diff
@@ -3,26 +3,29 @@ package x509util
...
 	"crypto/rand"
 	"fmt"
-	"math"
 	"math/big"
 )
 
 var (
-	maxUint64 = new(big.Int).SetUint64(math.MaxUint64)
-	one       = big.NewInt(1)
+	maxUint128, ok = new(big.Int).SetString("340282366920938463463374607431768211455", 10) //(2^128 − 1)
```
Comment (azdagron) [inline pkg/common/x509util/serialnumber.go]: albeit unlikely such a bug would make it out to pruduction, we should probably panic early if this is fails to parse:

```
maxUint128 = mustSetString("340282366920938463463374607431768211455", 10) //(2^128 − 1)
```

**Pull request merged**